src/main/java/org/checkerframework/specimin/Slicer.java

```diff
@@ -0,0 +1,212 @@
+package org.checkerframework.specimin;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.body.TypeDeclaration;
+import com.github.javaparser.ast.stmt.BlockStmt;
+import com.github.javaparser.resolution.Resolvable;
+import com.github.javaparser.resolution.UnsolvedSymbolException;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.checkerframework.specimin.unsolved.AddInformationResult;
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolAlternates;
+import org.checkerframework.specimin.unsolved.UnsolvedSymbolGenerator;
+
+public class Slicer {
+  /**
+   * The slice of nodes.
+   *
+   * <p>We cannot use HashSet<Node> here; when nodes' children are modified, {@code slice.contains}
+   * no longer returns true, even if {@code equals} evaluates to true, {@code hashCode} returns the
+   * same value, and {@code ==} yields the same reference.
+   *
+   * <p>An {@code IdentityHashMap} is safe to use because we use the same compilation units
+   * everywhere, so all corresponding nodes will have the same reference.
+   */
+  private final Set<Node> slice = Collections.newSetFromMap(new IdentityHashMap<>());
+
+  /** The slice of generated symbols. */
+  private final Set<UnsolvedSymbolAlternates<?>> generatedSymbolSlice = new HashSet<>();
+
+  /** The Nodes that need to be processed in the main algorithm. */
+  private final Deque<Node> worklist = new ArrayDeque<>();
+
+  private final Set<Node> postProcessingWorklist = new HashSet<>();
+
+  private final Map<String, Path> existingClassesToFilePath;
+  private final String rootDirectory;
+  private final Map<String, CompilationUnit> toSlice;
+
+  private final UnsolvedSymbolGenerator unsolvedSymbolGenerator;
+  private final TypeRuleDependencyMap typeRuleDependencyMap;
+
+  public Slicer(
+      Map<String, Path> existingClassesToFilePath,
+      String rootDirectory,
+      Map<String, CompilationUnit> toSlice,
+      TypeRuleDependencyMap typeRuleDependencyMap,
+      UnsolvedSymbolGenerator unsolvedSymbolGenerator) {
+    this.existingClassesToFilePath = existingClassesToFilePath;
+    this.rootDirectory = rootDirectory;
+    this.toSlice = toSlice;
+    this.typeRuleDependencyMap = typeRuleDependencyMap;
+    this.unsolvedSymbolGenerator = unsolvedSymbolGenerator;
+  }
+
+  public void slice() throws IOException {
```

Comment:
Is this method _always_ supposed to be called on a member of this class (I suspect so). If that's the case, make the contract simpler: this method can be private, and the constructor can call it.

Alternatively, make the constructor private and provide a static method that both constructs the object and runs the slice() method, and then returns the result to the caller.